crates/ruff_linter/src/noqa.rs

```diff
@@ -574,97 +713,81 @@ fn add_noqa_inner(
-                            .0
-                            .insert(rule);
+                        comments_by_line.push(Some(NoqaComment {
+                            line: directive_line.start(),
+                            diagnostic,
+                            directive: Some(directive.clone()),
+                        }));
                     }
                     continue;
                 }
             }
         }
 
         // There's no existing noqa directive that suppresses the diagnostic.
-        matches_by_line
-            .entry(locator.line_start(noqa_offset))
-            .or_insert_with(|| (RuleSet::default(), None))
-            .0
-            .insert(diagnostic.kind.rule());
+        comments_by_line.push(Some(NoqaComment {
+            line: locator.line_start(noqa_offset),
+            diagnostic,
+            directive: None,
+        }));
     }
 
-    let mut count = 0;
-    let mut output = String::with_capacity(locator.len());
-    let mut prev_end = TextSize::default();
-
-    for (offset, (rules, directive)) in matches_by_line {
-        output.push_str(locator.slice(TextRange::new(prev_end, offset)));
-
-        let line = locator.full_line(offset);
-
-        match directive {
-            None => {
-                // Add existing content.
-                output.push_str(line.trim_end());
-
-                // Add `noqa` directive.
-                output.push_str("  # noqa: ");
-
-                // Add codes.
-                push_codes(&mut output, rules.iter().map(|rule| rule.noqa_code()));
-                output.push_str(&line_ending);
-                count += 1;
-            }
-            Some(Directive::All(_)) => {
-                // Does not get inserted into the map.
-            }
-            Some(Directive::Codes(codes)) => {
-                // Reconstruct the line based on the preserved rule codes.
-                // This enables us to tally the number of edits.
-                let output_start = output.len();
-
-                // Add existing content.
-                output.push_str(
-                    locator
-                        .slice(TextRange::new(offset, codes.start()))
-                        .trim_end(),
-                );
-
-                // Add `noqa` directive.
-                output.push_str("  # noqa: ");
-
-                // Add codes.
-                push_codes(
-                    &mut output,
-                    rules
-                        .iter()
-                        .map(|rule| rule.noqa_code().to_string())
-                        .chain(codes.iter().map(ToString::to_string))
-                        .sorted_unstable(),
-                );
-
-                // Only count if the new line is an actual edit.
-                if &output[output_start..] != line.trim_end() {
-                    count += 1;
-                }
+    comments_by_line
+}
 
-                output.push_str(&line_ending);
-            }
+fn generate_noqa_edit(
```

Comment:
Should we make `NoqaEdit` a struct instead with two functions: 

* `write`: It writes the edit into a writer. This avoids allocating a string in the CLI use case
* `into_edit() -> Option<Edit>`: Returns an edit


I think this would help to closer match the performance of the CLI today where we allocate a single string to apply all edits (where this solution creates a string for each edit, and one for the merged output).

Reply:
I like this idea!